### DISCUSSION TITLE
Creating table in a pdf that spans multiple pages.

### DISCUSSION THREAD

**Discussion by aarthihWN** [pymupdf/PyMuPDF#2405] [Looking for help] [answered]
Hi,
My requirement is to create a table in pdf. It has many rows that makes it a long table that would span across multiple pages. I tried the recipe in the documentation [here](https://pymupdf.readthedocs.io/en/latest/recipes-stories.html#how-to-output-a-table). 

I've used the code as is except that the table_text has about 100 tuples. On executing, I find that the document ends up with only 2 pages, and many rows in between do not get written into the table. I'm unable to figure out if the problem is with the Story or DocumentWriter. Please help!

Thank you.

**Comment by JorjMcKie** [comment, marked as answer]
First you need to update to the latest version. You don't mention the one you are using, but earlier versions had their limitations.
Then I have a more complete, instructive example that covers 6 pages, repeats the header row and also gives rows alternating color backgrounds.
Would appreciate your feedback!
[national-capitals.zip](https://github.com/pymupdf/PyMuPDF/files/11463610/national-capitals.zip)

**Comment by aarthihWN** [comment]
Hi Jorj,
I am using 1.21.1 version. I had a look at your code and the resulting pdf. It is just what I need. The doc with all the styling and table headers, looks good. You just saved me a lot of effort with your ready made code. Thank you for this! Shall try it out with the latest version and get back to you if I face issues.
The code that generates the pdf seems to be the same. I am assuming that the version had some limitations. Thanks again!

**Reply by JorjMcKie** [reply]
Best of success!
There still do exist some limitations with supporting the HTML table element. E.g. automatic choice of column widths (cannot be influenced), multi-line table rows will never be split across pages, table header repeat is app code business - not automatic, no support of column or row groups.
This is an uncomplete list, but I think that a great deal can be achieved - presumably most of the usual requirements.

**Reply by aarthihWN** [reply]
Hi Jorj,
I was able to modify the table code to suit my requirements. Like you mentioned, I did have problems with fixing column width. The css styling was not taking effect. I used 'white-space: nowrap;' to get the format I required. So thanks again.

I have a question on the Xml.add_link. I used this function instead of add_text for one of the 'td' text. The text color changed to blue and also got the underline. But it didn't get translated to the hyperlink inside the pdf upon saving. It shows up as normal text and not as a link. Am I setting something incorrectly or this functionality is not available yet?

**Reply by JorjMcKie** [reply]
> Hi Jorj, I was able to modify the table code to suit my requirements. Like you mentioned, I did have problems with fixing column width. The css styling was not taking effect. I used 'white-space: nowrap;' to get the format I required. So thanks again.
> 
> I have a question on the Xml.add_link. I used this function instead of add_text for one of the 'td' text. The text color changed to blue and also got the underline. But it didn't get translated to the hyperlink inside the pdf upon saving. It shows up as normal text and not as a link. Am I setting something incorrectly or this functionality is not available yet?

Hm, following up on this. Don't yet know what is happening. It **_should_** work ...

**Reply by aarthihWN** [reply]
Hi Jorj, did you have a chance to look into the Xml to pdf hyperlinking?

**Reply by JorjMcKie** [reply]
> Hi Jorj, did you have a chance to look into the Xml to pdf hyperlinking?

I did. Still have no clou and submitted this to colleagues handling the translation html to pdf.

**Reply by JorjMcKie** [reply]
> Hi Jorj, did you have a chance to look into the Xml to pdf hyperlinking?

Me again.
Talked to my colleagues:
You have to use method `Story.write_with_links()` in order to support automatic hyperlink creation.
The syntax generated by `Xml.add_link()` is correct - that's not the problem.

We will adjust the documentation accordingly. Probably, this will lead to the recommendation to **_always_** use this `Story` writing alternative, because it may not always be the programmer who adds the link via Python Xml API: the same type of issue also happens if you pass HTML source to `Story` that contains `<a>` tags.
